toolkit/tools/imagegen/diskutils/diskutils.go

```diff
@@ -660,3 +721,92 @@ func getPartUUID(device string) (uuid string, err error) {
 	uuid = strings.TrimSpace(stdout)
 	return
 }
+
+func getSectorSizeFromFile(sectorFile string) (sectorSize uint64, err error) {
+	if exists, ferr := file.PathExists(sectorFile); ferr != nil {
+		logger.Log.Errorf("Error accessing sector size file %s", sectorFile)
+		err = ferr
+		return
+	} else if !exists {
+		err = fmt.Errorf("could not find the hw sector size file %s to obtain the sector size of the system", sectorFile)
+		return
+	}
+
+	fileContent, err := file.ReadLines(sectorFile)
+	if err != nil {
+		logger.Log.Errorf("Failed to read from %s: %s", sectorFile, err)
+		return
+	}
+
+	// sector file should only have one line, return error if not
+	if len(fileContent) != 1 {
+		err = fmt.Errorf("%s has more than one line", sectorFile)
+		return
+	}
+
+	sectorSize, err = strconv.ParseUint(fileContent[0], 10, 64)
+	return
+}
+
+func getSectorSize(diskDevPath string) (logicalSectorSize, physicalSectorSize uint64, err error) {
+	// Grab the specific disk name from /dev/xxx
+	diskName := diskDevPath[5:len(diskDevPath)]
+
+	hw_sector_size_file := fmt.Sprintf("/sys/block/%s/queue/hw_sector_size", diskName)
+	physical_sector_size_file := fmt.Sprintf("/sys/block/%s/queue/physical_block_size", diskName)
+
+	logicalSectorSize, err = getSectorSizeFromFile(hw_sector_size_file)
+	if err != nil {
+		return
+	}
+
+	physicalSectorSize, err = getSectorSizeFromFile(physical_sector_size_file)
+	return
+}
+
+func alignSectorAddress(sectorAddr, logicalSectorSize, physicalSectorSize uint64) (alignedSector uint64) {
+	// Need to make sure that starting sector of a partition is aligned based on the physical sector size of the system.
+	// For example, suppose the physical sector size is 4096. If the input start sector is 40960001, then this is misaligned,
+	// and need to be elevated to the next aligned address, which is (40960001/4096 + 1)*4096 = 4100096.
+
+	// We do need to take care of a special case, which is the first partition (normally boot partition) might be less than
+	// the physical sector size. In this case, we need to check whether the start sector is a multiple of 1 MiB.
+	if sectorAddr < physicalSectorSize {
+		if sectorAddr%(MiB/logicalSectorSize) == 0 {
+			alignedSector = sectorAddr
+		}
```

Comment:
Double-checking: if `sectorAddr < physicalSectorSize` is true but `sectorAddr%(MiB/logicalSectorSize) == 0` is false, is it expected for the function to return 0? If so, I'd add `alignedSector = 0` before this set of checks to make it clear that this is the value we're OK with. Also, I'd discourage depending on default values, since they might not always be clear to everyone and the language may change in the future. We're forced to do that with maps simulating sets but otherwise, let's be explicit about the values.

Reply:
Yes, when the condition errors out, then the function should return 0.

I'm not very clear about what you mean by being explicit about the values here. Which specific value are you referring to? Do you mean by the usage of "MiB" in the condition check?

Reply:
No, I meant `alignedSector`, since we've had a path where we're returning the default value for `int64`.